### PR TITLE
Support for `install:` in prepare

### PR DESCRIPTION
**Pull request by pvalena**
also better script handling.

## TODO:
 - [x] docs about koji and brew handling, maybe reword.
 - [x] ~~commandline args for `koji` or `brew`~~ - covered by install
 - [x] input validator
 - [x] more examples for install

### CODE REVIEW

**Comment by pvalena** [comment]
@psss WDYT, for a quick overview?

**Comment by pvalena** [comment]
[test]

File: spec/steps/prepare.fmf [hunk 1]
```diff
@@ -22,6 +22,18 @@ description: |
             how: shell
             script: dnf install -y httpd
 
+/install:
+    summary:
+        Install a package using dnf/yum.
+    description:
+        Installs a package from specified source and verifies that it was
+        installed successfuly.
```
Comment (psss) [inline spec/steps/prepare.fmf]: So the plan for this method is to support installing from all these sources?

 * default enabled repositories
 * copr
 * koji/brew

If so, could you please include examples for each source?

Comment (pvalena) [reply spec/steps/prepare.fmf]: Yes, that's in the TODO. :)

Comment (pvalena) [reply spec/steps/prepare.fmf]: What are `default enabled repositories`?

Comment (psss) [reply spec/steps/prepare.fmf]: By `default enabled` I meant basically doing `dnf install -y $packages` (no `--enablerepo`).

Comment (pvalena) [reply spec/steps/prepare.fmf]: Ok, that's correct then.

Comment (psss) [reply spec/steps/prepare.fmf]: This does not seem to be resolved. There is still only the `copr` example. It's important to have at least one example for each method.

Comment (pvalena) [reply spec/steps/prepare.fmf]: It's on the TODO (see the description). Please don't make me track it in conversations.

**Comment by pvalena** [comment]
> We need to have the expected functionality described at l

Yes. This PR was just for early feedback.

File: spec/steps/prepare.fmf (hunk 1 above)
Comment (psss) [reply spec/steps/prepare.fmf]: I see, I haven't realized this is a WIP.

Comment (pvalena) [reply spec/steps/prepare.fmf]: Sorry, I did not state it explicitly before. I'll make it `WIP` every time it is :) for new PRs.

**Comment by pvalena** [comment]
@psss should be all good-to-go.

File: spec/steps/prepare.fmf [hunk 2]
```diff
@@ -27,6 +27,44 @@ description: |
...
+
+/install-koji:
+    summary:
+        Download build package(s) and install them.
+    description:
+        Downloads build packages from koji (or brew) and install them using dnf/yum.
+    example: |
+        prepare:
+            how: install
+            url: https://koji.fedoraproject.org/koji/buildinfo?buildID=1465484
```
Comment (psss) [inline spec/steps/prepare.fmf:64]: The `url` sounds very generic to me. You can install rpm packages directly from and `url` as well. What about calling the key `build`? This would a bit more specific and works for both koji and brew. We could also support fetching builds by nvr or id in the future, what do you think?

File: tmt/cli.py [hunk 3]
```diff
@@ -263,6 +262,15 @@ def provision(context, **kwargs):
 @click.option(
     '-p', '--playbook', metavar='PLAYBOOK',
     help='Path or URI to ansible playbook to run.')
+@click.option(
+    '-g', '--package', metavar='PACKAGE',
```
Comment (psss) [inline tmt/cli.py:266]: Let's skip the short option for now. Once the step is converted into a dynamic plugin we can add a `-p` option insteady which would be more intuitive.

File: spec/steps/prepare.fmf (hunk 2 above)
Comment (lukaszachy) [reply spec/steps/prepare.fmf:64]: Could it be all three build/task/tag ? E.g current example would be `build = 1465484`. 
When I need exact nvr and if we are ok with automatic nvr/id detection it would be `build = foo-1.f31`.
At the moment 'tag' might be useful for module scratch builgs `tag = name_of_tag`
Task is similar to build but accepts only id.

I'm not saying everything should be possible from the beginning, just possibility is enough for now.

File: tmt/steps/prepare/__init__.py [hunk 4]
```diff
@@ -86,58 +91,131 @@ def how_generic(self, how, what):
...
-        ## TODO: remove this after run(shell=True) is in provision.prepare()
-        try:
-            self.plan.provision.prepare('shell', f"rpm -V {packages}")
-        except GeneralError:
-            self.plan.provision.prepare('shell', f"dnf install -y {packages}")
-        return
-        ## <
+        return self.command(f"rpm -V {packages} || sudo dnf install -y {packages}")
 
+    def command(self, command, logfile=True):
```
Comment (psss) [inline tmt/steps/prepare/__init__.py:183]: Is it really necessary to store the command in an extra script and sync to guest and only then run the command? Why not just execute using `ssh` command as we already do for example for `shell` preparation?

**Comment by psss** [comment]
@pvalena, please also rebase on the latest master. Thanks.

File: spec/steps/prepare.fmf (hunk 2 above)
Comment (pvalena) [reply spec/steps/prepare.fmf:64]: @psss sure, `build` is fine with me.

@lukaszachy it actually does parse the URI and gets just the build number from it. Then it uses `$ koji ...` or `$ brew`, based on the host name, so we'd have to take that into account too. Opting for simple "build url" at first seemed best to me (the options are limitless off course).
Would build URL be useful, as a start? (Note that blacklisting install packages is ortogonal to this.)

File: tmt/cli.py (hunk 3 above)
Comment (pvalena) [reply tmt/cli.py:266]: OK.

**Comment by pvalena** [comment]
> I'm missing L1 `require` gathering and installation (as promised in #174).
> 
> Implemented in [pvalena/pull/2](https://github.com/pvalena/tmt/pull/2)

I do not intend it to be the part of this PR. That change is in the `discover` step. They can be merged separately.

**Comment by pvalena** [comment]
> @pvalena, please also rebase on the latest master. Thanks.

Sure.

File: tmt/steps/prepare/__init__.py [hunk 5]
```diff
@@ -86,58 +91,131 @@ def how_generic(self, how, what):
...
+                    self.how_koji('koji', how, what)
+                    continue
+                if not re.search(r"^brew\.", domain) is None:
+                    self.how_koji('brew', how, what)
+                    continue
+
+            rest += wha
+        self.install(rest)
+
+    def how_koji(self, com, how, what):
```
[the remark this answers is not in the document]
Comment (pvalena) [reply tmt/steps/prepare/__init__.py:140]: Command, that should be use for downloading. It's not meant as an API, and I didn't really know what else to name it to.

So `com` is either `koji` or `brew`, based on the regex matching on line 130.

File: tmt/steps/prepare/__init__.py [hunk 6]
```diff
@@ -86,58 +91,131 @@ def how_generic(self, how, what):
...
+            handles various URIs or packages themselves
+        """
+        if not type(what) is list:
+            return self.install(what)
+
+        rest = []
+        for wha in what:
+            if self.is_uri(wha):
+                domain = self.get_uri(wha).netloc
+                if not re.search(r"^koji\.", domain) is None:
```
[the remark this answers is not in the document]
Comment (pvalena) [reply tmt/steps/prepare/__init__.py:130]: I wonder why I didn't use that. Could it give me false positives somehow?

File: spec/steps/prepare.fmf [hunk 7]
```diff
@@ -27,6 +27,44 @@ description: |
...
+        Installs a package(s) from specified source and verifies that it was
+        installed successfuly.
+    example: |
+        prepare:
+            how: install
+            package: ruby
+    implemented:
+        - /tmt/steps/prepare
+
+/install-copr:
```
[the remark this answers is not in the document]
Comment (pvalena) [reply spec/steps/prepare.fmf:43]: Sure, thanks. I didn't know how to handle this.

File: tmt/steps/prepare/__init__.py (hunk 4 above)
Comment (pvalena) [reply tmt/steps/prepare/__init__.py:183]: That was my first intention, but IIRC it proved to be very tricky and unreliable to quote the commands from users:
  https://github.com/psss/tmt/pull/159/files/2e317f5014ce6851b3b9e9d24f7477da41e8b718#diff-4476c8abdfef907fe037b9351ba16b5cR205

But maybe I didn't try the `shlex`, variant. And there's one more reason- but let me get back to you on Friday, when I test it.

**Comment by lukaszachy** [comment]
> > I'm missing L1 `require` gathering and installation (as promised in #174).
> > Implemented in [pvalena/pull/2](https://github.com/pvalena/tmt/pull/2)
> 
> I do not intend it to be the part of this PR. That change is in the `discover` step. They can be merged separately.

Fair enough, I'll create the PR for psss' repo once this one is merged.

**Comment by psss** [comment]
This one has been obsoleted by the `install` plugin in #218.
